rasa/core/validator.py

```diff
@@ -0,0 +1,178 @@
+
+        self.domain = domain
+        self.intents = intents
+        self.valid_intents = []
+        self.valid_utterances = []
+        self.stories = stories
+
+    @classmethod
+    def from_files(
+        cls, domain_file: Text, nlu_data: Text, story_data: Text
+    ) -> "Validator":
+        """Create an instance from the domain, nlu and story files."""
+
+        domain = Domain.load(domain_file)
+        loop = asyncio.new_event_loop()
+        stories = loop.run_until_complete(
+            StoryFileReader.read_from_folder(story_data, domain)
+        )
+        intents = load_data(nlu_data)
+        return cls(domain, intents, stories)
+
+    def _search(self, vector: List[Any], searched_value: Any) -> bool:
+        """Searches for a element in a vector."""
+
+        vector.append(searched_value)
+        count = 0
+        while searched_value != vector[count]:
+            count += 1
+        if count == len(vector) - 1:
+            return False
+        else:
+            return True
+
+    def verify_intents(self):
+        """Compares list of intents in domain with intents in NLU training data."""
+
+        domain_intents = []
+        nlu_data_intents = []
+
+        for intent in self.domain.intent_properties:
+            domain_intents.append(intent)
+
+        for intent in self.intents.intent_examples:
+            nlu_data_intents.append(intent.data["intent"])
+
+        for intent in domain_intents:
+            found = self._search(nlu_data_intents, intent)
+            if not found:
+                logger.error(
+                    "The intent '{}' is listed in the domain file, but "
+                    "is not found in the NLU training data.".format(intent)
+                )
+            else:
+                self.valid_intents.append(intent)
+
+        for intent in nlu_data_intents:
+            found = self._search(domain_intents, intent)
+            if not found:
+                logger.error(
+                    "The intent '{}' is in the NLU training data, but "
+                    "is not listed in the domain.".format(intent)
+                )
+
+    def verify_intents_in_stories(self):
+        """Checks intents used in stories.
+
+        Verifies if the intents used in the stories are valid, and whether
+        all valid intents are used in the stories."""
+
+        if self.valid_intents == []:
+            self.verify_intents()
+
+        stories_intents = []
+        for story in self.stories:
+            for event in story.events:
+                if type(event) == UserUttered:
+                    intent = event.intent["name"]
+                    stories_intents.append(intent)
+                    found = self._search(self.valid_intents, intent)
+
+                    if not found:
+                        logger.error(
+                            "The intent '{}' is used in stories, but is not a "
+                            "valid intent.".format(intent)
```

Comment:
What distinguishes a valid intent? This error showed up for me even though the intent was listed under intents in the domain file. I believe it had to do with the fact that this intent wasn't listed in the NLU data, but this shouldn't make it invalid.

Reply:
Looking at the code, looks like listed in domain but not NLU data is invalid, while listed in NLU data but not domain is valid. I think it should be the other way around (if it's not in the domain you definitely can't use it in stories!)